completion_aggregator/aggregator.py

```diff
@@ -0,0 +1,173 @@
+            else:
+                tick = nexttick
+                sleep_time = (tick - iteration_end_time).total_seconds()
+                time.sleep(sleep_time)
+
+        self.perform_cleanup()
+
+    def run_once(self):
+        """
+        Run the task once, then clean up and exit.
+        """
+        self.perform_service(timezone.now())
+        self.perform_cleanup()
+
+    def perform_service(self, tick):
+        """
+        Do the task we're here to do.
+        """
+        raise NotImplementedError
+
+    def perform_cleanup(self):
+        """
+        Perform a cleanup task if needed.
+        """
+        pass
+
+
+class AggregatorService(BasePeriodicService):
+    """
+    Enqueues tasks to reaggregate modified completions.
+
+    When blocks are completed, they mark themselves as stale.  This service
+    collects all stale blocks for each enrollment, and enqueues a single
+    recalculation of all aggregators containing those stale blocks.
+    """
+
+    service_name = "Aggregator Service"
+    enrollment_tuple = collections.namedtuple('enrollment_tuple', ['username', 'course_key'])
+
+    def perform_service(self, tick):
+        """
+        Collect stale blocks, and enqueue new aggregation tasks.
+        """
+        with transaction.atomic():
+            stale_queryset = models.StaleCompletion.objects.select_for_update().filter(resolved=False)
+            stale_blocks = collections.defaultdict(set)
+            forced_updates = set()
+            for stale in stale_queryset:
+                enrollment = self.enrollment_tuple(
+                    username=stale.username,
+                    course_key=stale.course_key,
+                )
+                if stale.block_key is None:
+                    stale_blocks[enrollment] = utils.BagOfHolding()
+                else:
+                    stale_blocks[enrollment].add(stale.block_key)
+                if stale.force:
+                    forced_updates.add(enrollment)
+
+            log.info("Performing aggregation update for %s user enrollments", len(stale_blocks))
+            for enrollment in stale_blocks:
+                if isinstance(stale_blocks[enrollment], utils.BagOfHolding):
+                    blocks = []
+                elif len(stale_blocks[enrollment]) > 128:
+                    # Limit the number of block_keys we will add to a task, because they will be sent over the network.
+                    blocks = []
+                else:
+                    blocks = [six.text_type(block_key) for block_key in stale_blocks[enrollment]]
+                tasks.update_aggregators.delay(
+                    username=enrollment.username,
+                    course_key=six.text_type(enrollment.course_key),
+                    block_keys=blocks,
+                    force=enrollment in forced_updates,
+                )
+            models.StaleCompletion.objects.filter(id__in=(stale.id for stale in stale_queryset)).update(resolved=True)
```

Comment:
Does work instead?
```python
stale_queryset.update(resolved=True)
```

Reply:
That was my initial thought, but I'm pretty sure it will re-evaluate the queryset, and cause race conditions.  Let me run a quick test....

Reply:
```python
>>>  qs.update(resolved=True)
2
>>> connection.queries[-1]
{'sql': 'UPDATE "completion_aggregator_stalecompletion" SET "resolved" = 1 WHERE "completion_aggregator_stalecompletion"."resolved" = 0',
 'time': '0.001'}
```
Which has the race condition (any new unresolved StaleCompletions will be resolved.  vs, the other version.

```python
 {'sql': 'UPDATE "completion_aggregator_stalecompletion" SET "resolved" = 1 WHERE "completion_aggregator_stalecompletion"."id" IN (3, 4)',
  'time': '0.000'}
```

This version explicitly asks for the blocks it actually handled.

Reply:
Ah yes!  You're right.  👍 

Can we get away without the `transaction.atomic` and `select_for_update` then?

Reply:
Hmm.  I'll need to investigate a little bit.  The main reason I had the select_for_update was to protect the rows against later queries, but aggregation is idempotent, and the table is only accessed in a create -> resolve -> delete pattern, so it should be safe not to protect it.  On the other hand, if the command is run as an infinite loop, we'll need transactions to be committed somewhere.  That could either be the way we have it, or more simply, using autocommit behavior (each query commits itself).  I'm not sure what the default behavior is, so I'll have to investigate.